### PR TITLE
Dreidenbach/dev

### PR DESCRIPTION
**Pull request by nvdreidenbach**
### Overview
- Created inference_perceiver.py as the internal model wrapper for Benchmarking the frozen perceiver MegaMolBART implementation
       - Updated data cache to work with a variable batch size and batched generation methods which resulted in the massive speed up for overall benchmarking
       - Added nbit specification for embeddings metrics to account for variability in user defined perceiver models
       - Added Non Identicality, Effective Novelty, Scaffold and Entropy Metrics
       - I made needed API changes to allow for benchmarking of various model's and behaviors (ie averaging vs non averaging fixed sized latent embeddings)

- Moved path setup to container launch instead of benchmarking start to allow for easier debugging and live development of the framework within the container. I added other mounts in the container launch to allow for importing of my custom model implementation

- Updated legacy CDDD code to work with current env and batched method of generation

The plan is to ensure the metrics are doing what is expected. Most of the streamlining and API changes are currently being handled via a fork of this update in @rilango branch. So a bulk of the main/cache/data usage will be overwritten in the next update which will not be dependent on my frozen model implementation.

### Future Changes in progress
- Cleaning of the Repo. I found it hard to find everything and there is a lot of code that is older or is not being used in the current Benchmarking workflow. 
- Confusing method names and double wrapper usage. This is already a WIP but having a wrapper for another wrapper along with the repo layout made it harder to navigate development.
- A future nice to have would be the ability to tag the generated models with all of their benchmarks. We already have is_valid so it would be nice to have is_unique or is_novel and so on so that the user can run the benchmarking framework get the numerical results and be able to easily query for their generated molecules of specified behavior.

### CODE REVIEW

File: benchmark/cuchembench/metrics/embedding.py [hunk 1]
```diff
@@ -176,15 +175,14 @@ def _calculate_metric(self, embeddings, fingerprints, top_k=None):
...
         return corr
 
-    def calculate(self, top_k=None, **kwargs):
+    def calculate(self, top_k=None, average_tokens = False, **kwargs):
 
         start_time = time.time()
         cache = Cache()
-        embeddings = cache.get_data('NN_embeddings')
-
+        embeddings = None #cache.get_data('NN_embeddings')
```
Comment (rilango) [inline benchmark/cuchembench/metrics/embedding.py]: Nit pick: Please remove this line and the condition below.

This fetch from cache was saving about 7 hrs per iteration. It accounted for saving about one week from Embedding metric. Please leave a "TODO" statement here so that I can revisit it later if we get into performance issues again.

**Review by rilango — APPROVED** [review]

File: benchmark.sh [hunk 2]
```diff
@@ -28,18 +28,23 @@ done
...
-export PYTHONPATH_CUCHEM="${PYTHONPATH_CUCHEM}:${WORKSPACE}/benchmark:${WORKSPACE}/cuchem:"
-
-export PYTHONPATH_MEGAMOLBART="${WORKSPACE}/common:${WORKSPACE}/common/generated:${WORKSPACE}/megamolbart"
-export PYTHONPATH_MEGAMOLBART="${PYTHONPATH_MEGAMOLBART}:${WORKSPACE}/benchmark:${WORKSPACE}/cuchem:"
-
-docker-compose \
-  -f ${SOURCE_ROOT}/setup/docker_compose.yml \
-  --project-directory ${SOURCE_ROOT} \
-  ${ACTION}
+# export CUCHEM_UI_START_CMD="python3 -m cuchembench --config-dir ${CONFIG_DIR}"
```
Comment (michalivne) [inline benchmark.sh]: If comments aren't needed, perhaps we should remove them?

File: benchmark/cuchembench/__main__.py [hunk 3]
```diff
@@ -138,22 +145,40 @@ def main(cfg):
...
 
     max_seq_len = int(cfg.sampling.max_seq_len)
+    exp_name = cfg.model.exp_name
     if cfg.model.name == 'MegaMolBART':
         from cuchembench.inference.megamolbart import MegaMolBARTWrapper
-        inferrer = MegaMolBARTWrapper()
+        inferrer = MegaMolBARTWrapper(checkpoint_file = cfg.model.checkpoint_file)
+        encoder_type = inferrer.megamolbart.model.model.encoder_type
+        if not cfg.model.perceiver_average and encoder_type == 'perceiver':
+            nbits = inferrer.megamolbart.model.model.max_seq_len * inferrer.megamolbart.model.model.steps_model #512 * k
```
Comment (michalivne) [inline benchmark/cuchembench/__main__.py:154]: Consider changing name? This is not number of bits but rather dimensions (`ndims`?)

File: benchmark/cuchembench/__main__.py [hunk 4]
```diff
@@ -138,22 +145,40 @@ def main(cfg):
     os.makedirs(output_dir, exist_ok=True)
 
     max_seq_len = int(cfg.sampling.max_seq_len)
+    exp_name = cfg.model.exp_name
     if cfg.model.name == 'MegaMolBART':
         from cuchembench.inference.megamolbart import MegaMolBARTWrapper
```
Comment (michalivne) [inline benchmark/cuchembench/__main__.py:150]: Why do we prefer to import here and not at the top?

File: benchmark/cuchembench/config/benchmark_metrics.yaml [hunk 5]
```diff
@@ -11,44 +11,83 @@ root:
...
-  #   - model.params.url (50052 for lb grpc)
-  name: CDDD
-  training_data: /data/db/cddd_train.sqlit
-
-  # name:  cuchembench.inference.GrpcMegaMolBARTWrapper
-  # training_data: /data/db/zinc_train.sqlite3
-  # params:
-  #   url: nginx:50052
+  name: MegaMolBART #Latent #CDDD
+  exp_name: prcvr_k4 #vae_k4 #mim_k1_salv # prcvr_k4 #seq2seq # mim_same_aug_k1 #
```
Comment (michalivne) [inline benchmark/cuchembench/config/benchmark_metrics.yaml:15]: We should store here comments that are specific to a particular set of experiments (this should be store in an example script)

File: megamolbart/megamolbart/inference_perceiver.py [hunk 6]
```diff
@@ -0,0 +1,619 @@
...
+    world_size=1,
+    global_rank=0,
+    local_rank=0,
+    tensor_model_parallel_size=1,
+    seed=1234,
+)
+logger = logging.getLogger(__name__)
+
+
+class MegaMolBART():
```
Comment (michalivne) [inline megamolbart/megamolbart/inference_perceiver.py:38]:
```
class MegaMolBART(object):
```
Subtle but important difference in some cases.
Also can you add class description in docstring?

File: megamolbart/megamolbart/inference_perceiver.py [hunk 7]
```diff
@@ -0,0 +1,619 @@
...
+        if isinstance(smiles, str):
+            smiles = [smiles]
+
+        tokens = self.tokenizer.tokenize(smiles, pad=True)
+
+        # # Append to tokens and mask if appropriate. No Padding for Perceiver
+        if pad_length and self.encoder_type != 'perceiver':
+            for i in range(len(tokens['original_tokens'])):
+                n_pad = pad_length - len(tokens['original_tokens'][i])
+                tokens['original_tokens'][i] += [self.tokenizer.pad_token] * n_pad
```
Comment (michalivne) [inline megamolbart/megamolbart/inference_perceiver.py:101]: This will not raise an error if `n_pad < 0`. Is that the desired behaviour?
Also why not using https://pytorch.org/docs/stable/generated/torch.nn.utils.rnn.pad_sequence.html for padding?

File: megamolbart/megamolbart/inference_perceiver.py [hunk 8]
```diff
@@ -0,0 +1,619 @@
...
+    # TODO as the inverse_transform is used via the sql query for generation this function should take in 
+    # either any sized embedding or a list of embeddings and process them via an appropriate batch size
+    # z = model.smiles2embedding(smiles) out_smiles = model.embedding2smiles(z)
+    def embedding2smiles(self, embeddings, mem_pad_mask, k=1, sanitize=True):
+        return self.inverse_transform([embeddings], mem_pad_mask, k, sanitize)[0]
+
+    def inverse_transform(self, embeddings, mem_pad_mask, k=1, sanitize=True):
+        smiles_interp_list = []
+        with torch.no_grad():
+            with concurrent.futures.ThreadPoolExecutor(max_workers=4) as executor:
```
Comment (michalivne) [inline megamolbart/megamolbart/inference_perceiver.py:149]: Why not doing batch decoding followed by parallel string parsing? 
Running multiple decoding in parallel might slow things down. Did you happen to test the effect on speed?

File: megamolbart/megamolbart/inference_perceiver.py [hunk 9]
```diff
@@ -0,0 +1,619 @@
+#!/usr/bin/env python3
+
```
Comment (michalivne) [inline megamolbart/megamolbart/inference_perceiver.py:2]: Can you add short docstrings with description of functions?
The use cases of some functions is not straight forward.

File: megamolbart/megamolbart/inference_perceiver.py [hunk 10]
```diff
@@ -0,0 +1,619 @@
...
+            emb = interpolated_emb[:, idx, :]
+            emb = emb.unsqueeze(1)
+            dims.append(tuple(emb.shape))
+            embeddings.append(emb)
+        generated_mols = [smiles1] + generated_mols + [smiles2]
+        embeddings = [embedding1] + embeddings + [embedding2]
+        dims = [tuple(embedding1.shape)] + dims + [tuple(embedding2.shape)]
+        return generated_mols, embeddings, combined_mask, dims
+
+    def find_similars_smiles_list(self,
```
Comment (michalivne) [inline megamolbart/megamolbart/inference_perceiver.py:211]: Docstring is required here

File: launch.sh [hunk 11]
```diff
@@ -191,16 +192,21 @@ dev() {
         DOCKER_CMD="${DOCKER_CMD} -e PYTHONPATH=${PYTHONPATH_CUCHEM}:/workspace/benchmark"
         DOCKER_CMD="${DOCKER_CMD} -w /workspace/cuchem/"
     elif [[ ${CONTAINER_OPTION} -eq 2 ]]; then
-        DOCKER_CMD="${DOCKER_CMD} -v ${CONTENT_PATH}/models/megamolbart_v0.1/:/models/megamolbart/"
+        # DOCKER_CMD="${DOCKER_CMD} -v ${CONTENT_PATH}/models/megamolbart_v0.1/:/models/megamolbart/"
```
Comment (michalivne) [inline launch.sh]: Do we need the commented lines?

File: benchmark/cuchembench/datasets/base.py [hunk 12]
```diff
@@ -28,12 +28,15 @@ def __init__(self,
...
+        # prop_data = pd.DataFrame(columns=properties_cols)
+        # prop_data.to_csv(self.prop_data_path, index=False)
+        # #TODO cols must be passed in
         fp_data_path = os.path.join(
             pathlib.Path(__file__).parent.parent.absolute(), 'csv_data', fp_filename)
-        if not os.path.exists(fp_data_path):
-            logger.info(f'Fingerprint path {fp_data_path} does not exist, generating temporary fingerprints file.')
-            temp_dir = '/tmp' #tempfile.TemporaryDirectory().name
-            fp_data_path = os.path.join(temp_dir, self.fp_filename)
+        # if not os.path.exists(fp_data_path):
```
Comment (michalivne) [inline benchmark/cuchembench/datasets/base.py:36]: Is the commented behaviour deprecated?

File: benchmark/cuchembench/inference/megamolbart.py [hunk 13]
```diff
@@ -41,51 +41,67 @@ def is_ready(self, timeout: int = 10) -> bool:
...
-                            dim: int,
-                            pad_mask):
+                            pad_mask,
+                            storage = None,
+                            batch_size = 100):
         '''
         Converts input embedding to SMILES.
         @param transform_spec: Input spec with embedding and mask.
         '''
         import torch
```
Comment (michalivne) [inline benchmark/cuchembench/inference/megamolbart.py:67]: Why not global import?
Also, if unused may delete or comment?

File: common/cuchemcommon/data/helper/chembldata.py [hunk 14]
```diff
@@ -41,15 +41,16 @@
...
-    if os.path.exists(setup_script):
-        logger.info('Triggering db setup...')
-        result = run(['bash', '-c',
-                        f'cd /opt/nvidia/cheminfomatics && {setup_script} setup'])
-        logger.info(f'Model download result: {result.stdout}')
-        logger.info(f'Model download result: {result.stderr}')
-        if result.returncode != 0:
-            raise Exception('Error downloading model')
+    pass
+    # setup_script = '/opt/nvidia/cheminfomatics/launch.sh'
```
Comment (michalivne) [inline common/cuchemcommon/data/helper/chembldata.py:45]: Was this code added where needed?
If so remove comment

File: cuchem/cuchem/utils/data_peddler.py [hunk 15]
```diff
@@ -24,9 +24,10 @@ def download_cddd_models():
     else:
         shutil.rmtree(os.path.join(target_dir, 'default_model'), ignore_errors=True)
 
-    download_script = '/opt/cddd/download_default_model.sh'
+    # download_script = '/opt/cddd/download_default_model.sh'
+    download_script = '/workspace/cddd/download_default_model.sh'
     if not os.path.exists(download_script):
-        download_script = '/tmp/download_default_model.sh'
+        download_script = '/workspace/cddd/download_default_model.sh'
```
Comment (michalivne) [inline cuchem/cuchem/utils/data_peddler.py:30]: Do we always assume we are in a container?

File: cuchem/cuchem/wf/generative/cddd.py [hunk 16]
```diff
@@ -92,18 +144,157 @@ def interpolate_smiles(self,
...
+            if force_unique:
+                interp_df = self.compute_unique_smiles(interp_df,
+                                                       self.inverse_transform,
+                                                       scaled_radius=scaled_radius)
+
+            result_df.append(interp_df)
+
+        return pd.concat(result_df)
+
+# import grpc
```
Comment (michalivne) [inline cuchem/cuchem/wf/generative/cddd.py:191]: What are the comments? If not required remove.

**Review by michalivne — COMMENTED** [review]
Great PR! See comments below. Generally speaking, need to remove non-required commented code, and to add docstrings to publicly exposed methods.

File: benchmark/cuchembench/__main__.py (hunk 4 above)
Comment (nvdreidenbach) [reply benchmark/cuchembench/__main__.py:150]: Depending on the container certain imports would not be available. There is a better design currently in the branch which removes all the grpc services

File: benchmark/cuchembench/__main__.py (hunk 3 above)
Comment (nvdreidenbach) [reply benchmark/cuchembench/__main__.py:154]: nbits is what rdkit uses so I wanted to keep the names consistent

File: benchmark/cuchembench/datasets/base.py (hunk 12 above)
Comment (nvdreidenbach) [reply benchmark/cuchembench/datasets/base.py:36]: The whole caching mechanism for fingerprints has been redesigned so we can ignore this file.

File: benchmark/cuchembench/inference/megamolbart.py (hunk 13 above)
Comment (nvdreidenbach) [reply benchmark/cuchembench/inference/megamolbart.py:67]: Same logic as other inside import. This should be fixed in clean up branch

File: common/cuchemcommon/data/helper/chembldata.py (hunk 14 above)
Comment (nvdreidenbach) [reply common/cuchemcommon/data/helper/chembldata.py:45]: Need for this data download code is removed in clean up branch.

File: cuchem/cuchem/utils/data_peddler.py (hunk 15 above)
Comment (nvdreidenbach) [reply cuchem/cuchem/utils/data_peddler.py:30]: Yes

File: cuchem/cuchem/wf/generative/cddd.py (hunk 16 above)
Comment (nvdreidenbach) [reply cuchem/cuchem/wf/generative/cddd.py:191]: CDDD is being removed so the comments will stay to maintain the grpc version

File: megamolbart/megamolbart/inference_perceiver.py (hunk 6 above)
Comment (nvdreidenbach) [reply megamolbart/megamolbart/inference_perceiver.py:38]: Class will be deprecated in new update so will add descriptions in stable future branch

File: benchmark/cuchembench/config/benchmark_metrics.yaml (hunk 5 above)
Comment (nvdreidenbach) [reply benchmark/cuchembench/config/benchmark_metrics.yaml:15]: Experiment logic changing in future branch so for now the comments were just for my reference
